src/recovery.rs

```diff
@@ -238,24 +255,50 @@ impl Recovery {
         let mut largest_newly_acked_pkt_num = 0;
         let mut largest_newly_acked_sent_time = now;
 
+        let max_rtt = cmp::max(self.latest_rtt, self.rtt());
+
         // Detect and mark acked packets, without removing them from the sent
         // packets list.
         for r in ranges.iter() {
-            let lowest_acked = r.start;
-            let largest_acked = r.end - 1;
+            let lowest_acked_in_block = r.start;
+            let largest_acked_in_block = r.end - 1;
 
             let unacked_iter = self.sent[epoch]
                 .iter_mut()
                 // Skip packets that precede the lowest acked packet in the block.
-                .skip_while(|p| p.pkt_num < lowest_acked)
+                .skip_while(|p| p.pkt_num < lowest_acked_in_block)
                 // Skip packets that follow the largest acked packet in the block.
-                .take_while(|p| p.pkt_num <= largest_acked)
+                .take_while(|p| p.pkt_num <= largest_acked_in_block)
                 // Skip packets that have already been acked or lost.
-                .filter(|p| p.time_acked.is_none() && p.time_lost.is_none());
+                .filter(|p| p.time_acked.is_none());
 
             for unacked in unacked_iter {
                 unacked.time_acked = Some(now);
 
+                // Check if acked packet was already declared lost.
+                if unacked.time_lost.is_some() {
+                    // Calculate new packet reordering threshold.
+                    let pkt_thresh =
+                        self.largest_acked_pkt[epoch] - unacked.pkt_num + 1;
+                    let pkt_thresh = cmp::min(MAX_PACKET_THRESHOLD, pkt_thresh);
+
+                    self.pkt_thresh = cmp::max(self.pkt_thresh, pkt_thresh);
+
+                    // Calculate new time reordering threshold.
+                    let loss_delay = max_rtt.mul_f64(self.time_thresh);
+                    if now.duration_since(unacked.time_sent) > loss_delay {
+                        // TODO: do time threshold update
+                        self.time_thresh = 5_f64 / 4_f64;
+                    }
+
+                    // Undo congestion window update.
+                    self.congestion_window = self.congestion_window_prev;
```

Comment:
perhaps,
`self.congestion_window = max(self.congestion_window, self.congestion_window_prev)`

Reply:
Yes, thank you.